Std/Data/List/Perm.lean

```diff
@@ -0,0 +1,852 @@
+
+theorem concat_perm (l : List α) (a : α) : concat l a ~ a :: l := by simp
+
+theorem Perm.length_eq {l₁ l₂ : List α} (p : l₁ ~ l₂) : length l₁ = length l₂ :=
+  p.rec
+    rfl
+    (fun _x l₁ l₂ _p r => by simp [r])
+    (fun _x _y l => by simp)
+    (fun _p₁ _p₂ r₁ r₂ => Eq.trans r₁ r₂)
+
+theorem Perm.eq_nil {l : List α} (p : l ~ []) : l = [] :=
+  eq_nil_of_length_eq_zero p.length_eq
+
+theorem Perm.nil_eq {l : List α} (p : [] ~ l) : [] = l :=
+  p.symm.eq_nil.symm
+
+@[simp]
+theorem perm_nil {l₁ : List α} : l₁ ~ [] ↔ l₁ = [] :=
+  ⟨fun p => p.eq_nil, fun e => e ▸ Perm.refl _⟩
+
+@[simp]
+theorem nil_perm {l₁ : List α} : [] ~ l₁ ↔ l₁ = [] :=
+  perm_comm.trans perm_nil
+
+theorem not_perm_nil_cons (x : α) (l : List α) : ¬[] ~ x :: l
+  | p => by injection p.symm.eq_nil
+
+@[simp]
+theorem reverse_perm : ∀ l : List α, reverse l ~ l
+  | [] => Perm.nil
+  | a :: l => by
+    rw [reverse_cons]
+    exact (perm_append_singleton _ _).trans ((reverse_perm l).cons a)
+
+theorem perm_cons_append_cons {l l₁ l₂ : List α} (a : α) (p : l ~ l₁ ++ l₂) :
+    a :: l ~ l₁ ++ a :: l₂ :=
+  (p.cons a).trans perm_middle.symm
+
+@[simp]
+theorem perm_replicate {n : Nat} {a : α} {l : List α} :
+    l ~ replicate n a ↔ l = replicate n a :=
+  ⟨fun p => eq_replicate.2
+    ⟨p.length_eq.trans <| length_replicate _ _, fun _b m => eq_of_mem_replicate <| p.subset m⟩,
+    fun h => h ▸ Perm.refl _⟩
+
+@[simp]
+theorem replicate_perm {n : Nat} {a : α} {l : List α} :
+    replicate n a ~ l ↔ replicate n a = l :=
+  (perm_comm.trans perm_replicate).trans eq_comm
+
+@[simp]
+theorem perm_singleton {a : α} {l : List α} : l ~ [a] ↔ l = [a] :=
+  @perm_replicate α 1 a l
+
+@[simp]
+theorem singleton_perm {a : α} {l : List α} : [a] ~ l ↔ [a] = l :=
+  @replicate_perm α 1 a l
+
+theorem Perm.eq_singleton {a : α} {l : List α} (p : l ~ [a]) : l = [a] :=
+  perm_singleton.1 p
+
+theorem Perm.singleton_eq {a : α} {l : List α} (p : [a] ~ l) : [a] = l :=
+  p.symm.eq_singleton.symm
+
+theorem singleton_perm_singleton {a b : α} : [a] ~ [b] ↔ a = b := by simp
+
+theorem perm_cons_erase [DecidableEq α] {a : α} {l : List α} (h : a ∈ l) : l ~ a :: l.erase a :=
+  let ⟨_l₁, _l₂, _, e₁, e₂⟩ := exists_erase_eq h
+  e₂.symm ▸ e₁.symm ▸ perm_middle
+
+@[elab_as_elim]
+theorem perm_induction_on
+    {P : (l₁ : List α) → (l₂ : List α) → l₁ ~ l₂ → Prop} {l₁ l₂ : List α} (p : l₁ ~ l₂)
+    (nil : P [] [] .nil)
+    (cons : ∀ x l₁ l₂, (h : l₁ ~ l₂) → P l₁ l₂ h → P (x :: l₁) (x :: l₂) (.cons x h))
+    (swap : ∀ x y l₁ l₂, (h : l₁ ~ l₂) → P l₁ l₂ h →
+      P (y :: x :: l₁) (x :: y :: l₂) (.trans (.swap x y _) (.cons _ (.cons _ h))))
+    (trans : ∀ l₁ l₂ l₃, (h₁ : l₁ ~ l₂) → (h₂ : l₂ ~ l₃) → P l₁ l₂ h₁ → P l₂ l₃ h₂ →
+      P l₁ l₃ (.trans h₁ h₂)) : P l₁ l₂ p :=
+  have P_refl l : P l l (.refl l) :=
+    List.recOn l nil fun x xs ih => cons x xs xs (Perm.refl xs) ih
+  Perm.recOn p nil cons (fun x y l => swap x y l l (Perm.refl l) (P_refl l)) @trans
```

Comment:
Move up?
```suggestion
@[elab_as_elim]
def Perm.recOnAux
    {motive : (l₁ : List α) → (l₂ : List α) → l₁ ~ l₂ → Sort _} {l₁ l₂ : List α} (p : l₁ ~ l₂)
    (nil : motive [] [] .nil)
    (cons : ∀ x l₁ l₂, (h : l₁ ~ l₂) → motive l₁ l₂ h → motive (x :: l₁) (x :: l₂) (.cons x h))
    (swap : ∀ x y l₁ l₂, (h : l₁ ~ l₂) → motive l₁ l₂ h →
      motive (y :: x :: l₁) (x :: y :: l₂) (.trans (.swap x y _) (.cons _ (.cons _ h))))
    (trans : ∀ l₁ l₂ l₃, (h₁ : l₁ ~ l₂) → (h₂ : l₂ ~ l₃) → motive l₁ l₂ h₁ → motive l₂ l₃ h₂ →
      motive l₁ l₃ (.trans h₁ h₂)) : motive l₁ l₂ p :=
  have motive_refl l : motive l l (.refl l) :=
    List.recOn l nil fun x xs ih => cons x xs xs (Perm.refl xs) ih
  Perm.recOn p nil cons (fun x y l => swap x y l l (Perm.refl l) (motive_refl l)) @trans
```

Reply:
Maybe keep both and define `perm_induction_on` in terms of `Perm.recOnAux`? I don't know how often `perm_induction_on` was used.

Reply:
It appears to not be used in mathlib. I'm thinking maybe `Perm.swap'_induction_on` to reflect how it is different from recOn.

(btw, your `Perm.recOnAux` doesn't seem to be a generalization, since `motive` has to be in `Prop` -- is this intentional?)

Reply:
Using `Sort _` was a reflex, `Prop` is more accurate. The idea is that custom recursors preferably have `.rec*` names to make them identifiable and indicate that they should be used with `induction _ using ..` instead of `apply` and such. The original disguises itself as a normal theorem.

How about this version? It more closely matches `Perm.recOn` and uses `swap'`.
```lean
@[elab_as_elim]
def Perm.recOn'
    {motive : (l₁ : List α) → (l₂ : List α) → l₁ ~ l₂ → Prop} {l₁ l₂ : List α} (p : l₁ ~ l₂)
    (nil : motive [] [] .nil)
    (cons : ∀ (x) {l₁ l₂}, (h : l₁ ~ l₂) → motive l₁ l₂ h → motive (x :: l₁) (x :: l₂) (.cons x h))
    (swap' : ∀ x y {l₁ l₂}, (h : l₁ ~ l₂) → motive l₁ l₂ h →
      motive (y :: x :: l₁) (x :: y :: l₂) (.swap' _ _ h))
    (trans : ∀ {l₁ l₂ l₃}, (h₁ : l₁ ~ l₂) → (h₂ : l₂ ~ l₃) → motive l₁ l₂ h₁ → motive l₂ l₃ h₂ →
      motive l₁ l₃ (.trans h₁ h₂)) : motive l₁ l₂ p :=
  have motive_refl l : motive l l (.refl l) :=
    List.recOn l nil fun x xs ih => cons x (Perm.refl xs) ih
  Perm.recOn p nil cons (fun x y l => swap' x y (Perm.refl l) (motive_refl l)) @trans
```

Reply:
I like `Perm.recOnSwap'` if that sounds good to everyone.

Note that there isn't really a pattern established yet in Std. The other `elab_as_elim` recursors in Std:
```
Fin.succRec
Fin.succRecOn
Fin.induction
Fin.inductionOn
Fin.cases
Fin.reverseInduction
Fin.lastCases 
Fin.addCases
Nat.gcd.induction
Nat.recAux
Nat.recAuxOn
Nat.casesAuxOn
Nat.strongRec
Nat.strongRecOn
Nat.recDiagAux
Nat.recDiag
Nat.recDiagOn
Nat.casesDiagOn
WF.recC (private)
```
Presumably the ones with `induction` in the name were inherited from Mathlib? But I'm not convinced that `rec` is the most discoverable name for lemmas intended to be used in `induction using ...`

Reply:
The `Nat.gcd.induction` is the only slightly strange one I see. They all have rec, cases, ind, or induction in there, same as in core.

Reply:
They should also all be defs and not theorems, except for ind/induction ones which can go one way or the other. But do include a docstring even if it's a theorem.

Reply:
Is the difference between `recOn` and `inductionOn` just `Prop` vs `Sort u` as the motive?

Reply:
Yes, but it's mostly used when the `Prop` version is significantly different (e.g. quotient types where propositional irrelevance simplifies things quite a bit).

Reply:
PS: `Perm.recOnSwap'` sounds good to me!